### DISCUSSION TITLE
Pytrees: marking some containers as leaves.

### DISCUSSION THREAD

**Discussion by grahamgower** [google/flax#1717] [Q&A] [answered]
Pytees are a great idea, and I'd like to make use of them more. In particular, I want to use pytrees where the leaves are numpy (or jax.numpy) arrays. In general, this works great, but there's one usage that needs some care: `b = tree_map(np.shape, a)`. This results in `tree_structure(a) != tree_structure(b)`, which means that `b` can't be used to create new pytrees with the same array shapes. To work around this I made a simple np.shape wrapper so that the shape tuple isn't recognised as an internal pytree node (see code below). But I wonder if this is fragile, in the sense that future jax may choose to register new kinds of pytree nodes (e.g. collections.abc.Sequence), thus breaking the opaqueness of my wrapper. In particular, I've seen discussions about making dataclasses pytree nodes, so I wanted to double check that I'm not doing something silly here.

```python
import collections

import jax
import numpy as np
import pytest

class _OpaqueSequence(collections.abc.Sequence):
    def __init__(self, t):
        self._t = t

    def __len__(self):
        return len(self._t)

    def __getitem__(self, index):
        if not isinstance(index, int):
            raise TypeError("Sequence index must be an integer.")
        return self._t[index]

def shape(a: np.ndarray) -> collections.abc.Sequence:
    return _OpaqueSequence(a.shape)

@pytest.mark.parametrize(
    "a",
    [
        np.zeros((1, 2, 3)),
        {"x": np.zeros((1, 2, 3)), "y": np.zeros((4, 5))},
        {"x": {"y": np.zeros((1, 2, 3)), "w": {"z": np.zeros((4, 5))}}},
    ],
)
def test_shape(a):
    a_shape = jax.tree_map(shape, a)
    b = jax.tree_map(np.zeros, a_shape)
    np.testing.assert_equal(a, b)
```

**Comment by cgarciae** [comment, marked as answer]
Depending on the general structure of `a` you could alternatively use `is_leaf` like this:

```python
a_shape = jax.tree_map(lambda x: x.shape, a)
b = jax.tree_map(np.zeros, a_shape, is_leaf=lambda x: isinstance(x, tuple))
np.testing.assert_equal(a, b)
```

The obvious restriction here is that all tuples are expected to be shapes (`Tuple[int, ...]`), else your solution is not bad since custom classes are not registered by default as pytrees with the exception of `NamedTuple`s and `dataclass`.

**Reply by grahamgower** [reply]
Thanks, somehow I'd missed the `is_leaf` argument. That's very useful in this kind of scenario! In any event, I just realised that my earlier solution is largely unnecessary, as the shape can just be converted to a numpy array: `a_shape = jax.tree_map(lambda x: np.array(x.shape), a) `, and numpy arrays are certainly not treated as pytree nodes.
